app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/SenderDeviceAsyncTask.java

```diff
@@ -0,0 +1,112 @@
+import android.widget.Toast;
+
+import org.kiwix.kiwixmobile.BuildConfig;
+import org.kiwix.kiwixmobile.R;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.DeviceListFragment.FILE_TRANSFER_PORT;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.DeviceListFragment.getFileName;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.showToast;
+
+/**
+ * Helper class for the local file sharing module, used in {@link DeviceListFragment}.
+ *
+ * Once the handshake between the two connected devices has taked place, this async-task is used
+ * on the sender device to transfer the file to the receiver device at the FILE_TRANSFER_PORT port.
+ *
+ * It takes in the uri of a single file, and copies all the bytes from input stream of the file to
+ * the output stream of the receiver device. Also changes the status of the corresponding FileItem
+ * on the list of files for transfer in {@link TransferProgressFragment}.
+ *
+ * A new task is created by the sender for every file to be transferred
+ * */
+class SenderDeviceAsyncTask extends AsyncTask<Uri, Void, Boolean> {
+
+  private static final String TAG = "SenderDeviceAsyncTask";
+
+  private DeviceListFragment deviceListFragment;
+  private TransferProgressFragment transferProgressFragment;
+  private int fileItemIndex;
+
+  public SenderDeviceAsyncTask(DeviceListFragment deviceListFragment, TransferProgressFragment transferProgressFragment, int fileItemIndex) {
+    this.deviceListFragment = deviceListFragment;
+    this.transferProgressFragment = transferProgressFragment;
+    this.fileItemIndex = fileItemIndex;
+  }
+
+  @Override
+  protected void onPreExecute() {
+    transferProgressFragment.changeStatus(fileItemIndex, FileItem.SENDING);
+  }
+
+  @Override
+  protected Boolean doInBackground(Uri... fileUris) {
+    Uri fileUri = fileUris[0];    // Uri of file to be transferred
+    Socket socket = new Socket(); // Represents the sender device
+    boolean fileSendSuccessful;   // Whether this task was successful in sending the file
+
+    try {
+      socket.bind(null);
+
+      String hostAddress = deviceListFragment.getFileReceiverDeviceAddress().getHostAddress();
+      socket.connect((new InetSocketAddress(hostAddress, FILE_TRANSFER_PORT)), 15000);
+
+      if(BuildConfig.DEBUG) Log.d(TAG, "Sender socket - " + socket.isConnected());
+
+      OutputStream socketOutputStream = socket.getOutputStream();
+
+      ContentResolver contentResolver = deviceListFragment.getActivity().getContentResolver();
+      InputStream fileInputStream = contentResolver.openInputStream(fileUri);
+
+      DeviceListFragment.copyToOutputStream(fileInputStream, socketOutputStream);
+      if(BuildConfig.DEBUG) Log.d(TAG, "Sender: Data written");
+
+      fileSendSuccessful = true;
```

Comment:
you don't need this variable, just return at these sites where you set it

Reply:
There's a `finally` section which will have to be repeated in both try and catch, plus added benefit of readability for what is being sent to onPostExecute()

Reply:
Hmm?
The finally executes after the return. The best advantage of `finally`

onPostExecute is already pretty clear about what it is about.
`protected void onPostExecute(Boolean fileSendSuccessful) {`

returning values at the correct location with the appropriate context gives more information. Introducing a variable requires looking at the assignments of the variable to truly know what is happening.

Reply:
`finally` really does mean _finally_ lol
Thanks for this, I'll get to it

Reply:
it does and in super funky ways
`try{return true} finally{return false}`
will return false. Don't think about it too much and also never ever do that

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/59bb020e9173f5a0e1b729256cac69b6acb6ce84#diff-0431b660d7990a5ca00c1b06f9912594R76)